Lib/test/test_dataclasses/__init__.py

```diff
@@ -4869,5 +4869,125 @@ class A:
+
+            def _set_foo(slf, value):
+                self.assertIs(__class__, type(slf))
+
+            def _del_foo(slf):
+                self.assertIs(__class__, type(slf))
+
+            foo = property(_get_foo, _set_foo, _del_foo)
+
+        a = A()
+        self.assertIs(a.foo, A)
+        a.foo = 4
+        del a.foo
+
+    def test_zero_argument_super_with_new_property(self):
+        @dataclass(slots=True)
+        class A:
+            @property
+            def foo(slf):
+                return slf.__class__
+
+            @foo.setter
+            def foo(slf, value):
+                self.assertIs(__class__, type(slf))
+
+            @foo.deleter
+            def foo(slf):
+                self.assertIs(__class__, type(slf))
+
+        a = A()
+        self.assertIs(a.foo, A)
+        a.foo = 4
+        del a.foo
+
+    # Test the parts of a property individually.
+    def test_slots_dunder_class_property_getter(self):
+        @dataclass(slots=True)
+        class A:
+            @property
+            def foo(slf):
+                return __class__
+
+        a = A()
+        self.assertIs(a.foo, A)
+
+    def test_slots_dunder_class_property_setter(self):
+        @dataclass(slots=True)
+        class A:
+            foo = property()
+            @foo.setter
+            def foo(slf, val):
+                self.assertIs(__class__, type(slf))
+
+        a = A()
+        a.foo = 4
+
+    def test_slots_dunder_class_property_deleter(self):
+        @dataclass(slots=True)
+        class A:
+            foo = property()
+            @foo.deleter
+            def foo(slf):
+                self.assertIs(__class__, type(slf))
+
+        a = A()
+        del a.foo
+
+    def test_wrapped(self):
+        def mydecorator(f):
+            @wraps(f)
+            def wrapper(*args, **kwargs):
+                return f(*args, **kwargs)
+            return wrapper
+
+        @dataclass(slots=True)
+        class A:
+            @mydecorator
+            def foo(self):
+                super()
+
+        A().foo()
+
+    def test_remembered_class(self):
+        # Apply the dataclass decorator manually (not when the class
+        # is created), so that we can keep a reference to the
+        # undecorated class.
+        class A:
+            def cls(self):
+                return __class__
+
+        self.assertIs(A().cls(), A)
+
+        B = dataclass(slots=True)(A)
+        self.assertIs(B().cls(), B)
+
+        # This is probably undesirable behavior, but is a function of
+        # how modifying __class__ in the closure works.  I'm not sure
+        # this should be tested or not: I don't really want to
+        # guarantee this behavior, but I don't want to lose the point
+        # that this is how it works.
+
+        # The underlying class is "broken" by changing its __class__
+        # in A.foo() to B.  This normally isn't a problem, because no
+        # one will be keeping a reference to the underlying class A.
```

Comment:
Another way to do this fix would be to actually replace the closure tuple, on each function we find, with a new closure tuple that actually includes the new _cell_, from the new class, rather than mutating the contents of the cell from the old class. This would avoid breaking the old class! But it would also have some downsides:
- creating a new closure tuple for each function we find would be slower.
- inaccessible functions (e.g. those hidden inside opaque decorators that don't use `functools.wraps`) wouldn't be fixed (as opposed to the current PR, where they are fixed as long as at least one accessible function is found.)

Probably the second downside is big enough to kill this idea.

Reply:
Yeah, I think so. Plus I don't think anyone actually uses code like this (famous last words, I realize). And it occurred to me that it's not like the dataclass decorator doesn't alter the original class, anyway: it already adds members and sets attributes. I'll grant you that's not quite in the same league as modifying the closures like this change does, but in general I think the rule should be that the decorator can step on the original class any way it likes.